Add support for GCP scopes

bgdanix opened the pull request:
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | yes
| 🔗 Related issues | #1284
| ❓ Documentation  | yes

## Description

Add support for specifying the search scopes when using GCP provider

sundowndev commented on pkg/remote/google/repository/cloudresourcemanager.go:
```diff
@@ -25,26 +28,40 @@ func NewCloudResourceManagerRepository(service *cloudresourcemanager.Service, co
+			erorsString = erorsString + "Project: " + project + " had the following error: " + errval.Error() + "; "
+		}
+		return bindingsByProject, errors.New(erorsString)
+	} else {
+		return bindingsByProject, nil
+	}
```
this else case is useless, you can just return at the end

bgdanix replied:
Agree, implemented

sundowndev commented on pkg/remote/google/repository/asset.go:
```diff
@@ -75,101 +74,111 @@ func NewAssetRepository(client *asset.Client, config config.GCPTerraformConfig,
-			if err != nil {
-				return nil, err
+	for _, scope := range s.config.Scope {
+		cacheKey := "SearchAllResources" + scope
```
Could you please use the same pattern as for other repositories ?

```go
cacheKey := fmt.Sprintf("SearchAllResources_%s", scope)
```

bgdanix replied:
Done

sundowndev reviewed (changes requested):
Thanks for the PR @bgdanix 👍🏻 

I left some comments about the implementation, I generally agree with the approach, I think code now lack of readability and testing.

Some issues I had while testing your code : 

- The identifier for the organization is wrongly asserted and lead to a rpc error.
- I get an unexpected result when one of the GCP scopes is having permission issues, it ignores listing of some resources for all scopes. This is expected from a technical POV, but not from a user POV. We may have UX concerns about this behavior.

Could you address the comments until we review this again ? It'd be nice to get a review from someone else on this one.

bgdanix commented on pkg/cmd/scan.go:
```diff
@@ -323,6 +345,26 @@ func scanRun(opts *pkg.ScanOptions) error {
 
+func parseScopeFlag(scope []string) ([]string, error) {
+
+	scopeRegex := `projects/\S*$|folders/\d*$|organizations/\S*$`
```
I've crafted this regex. The reason I did not put more constraints for organizations was to make future-proof, in case GCP will switch to a different style of ID. But thinking about it, I did put the constraint on the folder scope, I'll add it also to the org

bgdanix commented on pkg/remote/google/config/config.go:
```diff
@@ -1,7 +1,5 @@
-	Project string `cty:"project"`
-	Region  string `cty:"region"`
-	Zone    string `cty:"zone"`
+	Scope []string `cty:"scope"`
```
Implemented

bgdanix commented on pkg/remote/google/repository/cloudresourcemanager.go:
```diff
@@ -25,26 +28,40 @@ func NewCloudResourceManagerRepository(service *cloudresourcemanager.Service, co
-	}
+	bindingsByProject := make(map[string]map[string][]string)
+	errorsByProject := make(map[string]error)
+	var erorsString string
```
Implemented

bgdanix commented on pkg/cmd/scan.go:
```diff
@@ -124,6 +140,12 @@ func NewScanCmd(opts *pkg.ScanOptions) *cobra.Command {
 	)
+	fl.StringSliceP(
+		"gcp-scope",
+		"s",
```
Done

bgdanix commented:
>     * The identifier for the organization is wrongly asserted and lead to a rpc error.
> 
>     * I get an unexpected result when one of the GCP scopes is having permission issues, it ignores listing of some resources for all scopes. This is expected from a technical POV, but not from a user POV. We may have UX concerns about this behavior.

Thank you @sundowndev for taking time to review this. I have addressed all your comments. The above reported issues should also be fixed now.

sundowndev commented on pkg/cmd/scan.go:
```diff
@@ -142,7 +142,7 @@ func NewScanCmd(opts *pkg.ScanOptions) *cobra.Command {
 	fl.StringSliceP(
 		"gcp-scope",
-		"s",
+		"",
```
Please use `fl.StringSlice()` instead so you don't need this empty string

bgdanix replied:
Ah, yeah, makes sense. Done

sundowndev commented on pkg/remote/google/repository/asset.go:
```diff
@@ -76,9 +78,10 @@ func NewAssetRepository(client *asset.Client, config config.GCPTerraformConfig,
 func (s assetRepository) listAllResources(ty string) ([]*assetpb.Asset, error) {
 
 	filteredResults := []*assetpb.Asset{}
+	var erorString string
```
errorString*

bgdanix replied:
Missed it. Now should be ok

sundowndev commented on pkg/remote/google/repository/asset_test.go:
```diff
@@ -52,10 +52,10 @@ func Test_assetRepository_searchAllResources_CacheMiss(t *testing.T) {
-	c.On("Unlock", "SearchAllResources").Times(1)
-	c.On("Put", "SearchAllResources", mock.IsType([]*assetpb.ResourceSearchResult{})).Return(false).Times(1)
-	repo := NewAssetRepository(assetClient, config.GCPTerraformConfig{Scope: []string{""}}, c)
+	c.On("GetAndLock", "SearchAllResources_").Return(nil).Times(1)
+	c.On("Unlock", "SearchAllResources_").Times(1)
+	c.On("Put", "SearchAllResources_", mock.IsType([]*assetpb.ResourceSearchResult{})).Return(false).Times(1)
+	repo := NewAssetRepository(assetClient, config.GCPTerraformConfig{Scopes: []string{""}}, c)
```
Is there a particular reason to use an empty scope here ? I find it confusing to read

bgdanix replied:
The only reason I kept it empty was because previously it also had empty project, region, etc. Now I've added a dummy scope

sundowndev commented on pkg/remote/google/repository/cloudresourcemanager.go:
```diff
@@ -58,10 +59,10 @@ func (s *cloudResourceManagerRepository) ListProjectsBindings() (map[string]map[
 	if len(errorsByProject) > 0 {
 		for project, errval := range errorsByProject {
-			erorsString = erorsString + "Project: " + project + " had the following error: " + errval.Error() + "; "
+			erorString = erorString + fmt.Sprintf("Project: %s had the following error: %s; ", project, errval.Error())
```
Not sure to get the end of the template `%s; `, the semicolon is unnecessary here

bgdanix replied:
The reason I've put the semicolon is to separate between multiple errors. E.g. 
`For scope projects/123456 on resource TheResource got error: rpc error: code = 123 desc = description; For scope projects/123456 on resource AnotherResource got error: abc`

eliecharra commented on pkg/remote/google/repository/asset.go:
```diff
@@ -75,102 +76,131 @@ func NewAssetRepository(client *asset.Client, config config.GCPTerraformConfig,
+		defer s.cache.Unlock(cacheKey)
+
+		req := &assetpb.ListAssetsRequest{
+			Parent:      scope,
```
I'm not sure that a folder is supported for `listAssets` call according to the documentation here : https://cloud.google.com/asset-inventory/docs/reference/rpc/google.cloud.asset.v1#google.cloud.asset.v1.ListAssetsRequest

Can you double check that ?

bgdanix replied:
Indeed. Ideally we should not use listAssets but only searchAssets as the main difference is that list takes a snapshot of how the assets were at a particular time, which could result in a diff from what it's actually present.

eliecharra replied:
Where did you get this info ? I asked this question [here](https://stackoverflow.com/questions/69648122/difference-between-listassets-and-searchallresources) a couple of month ago and I didn't get any answer.

bgdanix replied:
From [here](https://cloud.google.com/asset-inventory/docs/listing-assets) there are a couple of hints about it:

```
Listing assets might not meet the performance requirements for large scale customers. If this is the case for you and calling list API encounters timeout, it's recommended to use export API instead.
```

Also in [the API](https://cloud.google.com/asset-inventory/docs/reference/rest/v1/assets) it's mentioned the purpose is to list resources from a specific timeframe

```
Lists assets with time and resource types and returns paged results in response.
readTime:  Timestamp to take an asset snapshot.  If not specified, the current time will be used. Due to delays in resource data collection and indexing, there is a volatile window during which running the same query may get different results.
```

While the search is more appropriate for our intent:

```
The Cloud Asset API allows you to use a custom query language to query resource metadata on a project, folder, or organization.
Method: searchAllResources. Searches all Cloud resources within the specified scope, such as a project, folder, or organization. 
```

eliecharra commented on pkg/remote/google/provider.go:
```diff
@@ -55,10 +53,8 @@ func (p *GCPTerraformProvider) Version() string {
+func (p *GCPTerraformProvider) SetConfig(scopes []string) config.GCPTerraformConfig {
 	return config.GCPTerraformConfig{
-		Project: os.Getenv("CLOUDSDK_CORE_PROJECT"),
-		Region:  os.Getenv("CLOUDSDK_COMPUTE_REGION"),
-		Zone:    os.Getenv("CLOUDSDK_COMPUTE_ZONE"),
```
Why do you removed Region and Zone parameters ?

bgdanix replied:
Because they are not needed since we use scopes.

eliecharra commented on pkg/remote/google/provider.go:
```diff
@@ -55,10 +53,8 @@ func (p *GCPTerraformProvider) Version() string {
-		Project: os.Getenv("CLOUDSDK_CORE_PROJECT"),
-		Region:  os.Getenv("CLOUDSDK_COMPUTE_REGION"),
-		Zone:    os.Getenv("CLOUDSDK_COMPUTE_ZONE"),
+		Scopes: scopes,
```
I think you missed the point, scopes are not what we are expecting here in this struct https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#scopes

I think this PR is more complicated than you think, it seems that it is not possible to initialize a `google` terraform provider with a folder or organization scope (not to be confused with OAuth scopes). You can only configure a project.

We probably need to create another kind of detail fetcher that will lazy instantiate one terraform providers per project. We are doing that for S3 buckets, but it's a really tricky thing.

Did you try to run this PR in deep mode ? I think this is broken

bgdanix replied:
The entire point I had in mind was to support multiple or boarder scopes for scanning. I did not run it in deep mode since it's not recommended, but for a simple mode targeting various projects or an entire GCP org it works really good for me:

```json
	"summary": {
		"total_resources": 400,
		"total_changed": 0,
		"total_unmanaged": 398,
		"total_missing": 1,
		"total_managed": 1
	}
```

bgdanix commented:
@sundowndev I've changed the code a bit to cater for cases when multiple scopes are used and one (or all) of them has permission error. Might not be the best way, but couldn't think of a better implementation for this case.

eliecharra commented on pkg/remote/google/provider.go:
```diff
@@ -55,10 +53,8 @@ func (p *GCPTerraformProvider) Version() string {
 }
 
-func (p *GCPTerraformProvider) GetConfig() config.GCPTerraformConfig {
+func (p *GCPTerraformProvider) SetConfig(scopes []string) config.GCPTerraformConfig {
```
`GetConfig` was ok, we are not setting anything in this function

bgdanix replied:
Well there's also nothing to get as during automated tests we have to set the config (scope)

eliecharra commented on pkg/cmd/scan.go:
```diff
@@ -62,6 +62,22 @@ func NewScanCmd(opts *pkg.ScanOptions) *cobra.Command {
+				if err != nil {
+					return err
+				}
+			} else if to != common.RemoteGoogleTerraform && len(GCPScope) > 0 {
+				return errors.New("gcp-scope can only be utilized when using " + common.RemoteGoogleTerraform + " flag")
+			} else if to == common.RemoteGoogleTerraform && len(GCPScope) == 0 {
+				return errors.New("gcp-scope must be specified when using " + common.RemoteGoogleTerraform + " flag")
+			}
```
Plea try to avoid else-if statements, there is not reason do use them here

eliecharra reviewed (changes requested):
This PR cannot work as it, terraform provider should be reworked. 
I will do another review pass later
